olp-cpp-sdk-core/include/olp/core/thread/SyncQueue.h

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (C) 2019 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+#pragma once
+
+#include <condition_variable>
+#include <deque>
+#include <mutex>
+#include <queue>
+
+#include <olp/core/CoreApi.h>
+
+namespace olp {
+namespace thread {
+
+/**
+ * @brief The SyncQueue class is a thread safe container adapter.
+ *
+ * @tparam T queue item to be stored inside the buffer.
+ * @tparam Container The type of the underlying container that should mimic the
+ * API of std::queue providing the following functions:
+ * - empty()
+ * - push(T&&)
+ * - push(const T&)
+ * - front()
+ * - pop()
+ */
+template <typename T, typename Container>
+class CORE_API SyncQueue final {
+ public:
+  SyncQueue() noexcept;
+  ~SyncQueue();
+
+  // Non-copyable, non-movable
+  SyncQueue(const SyncQueue&) = delete;
+  SyncQueue& operator=(const SyncQueue&) = delete;
+  SyncQueue(SyncQueue&&) = delete;
+  SyncQueue& operator=(SyncQueue&&) = delete;
+
+  /**
+   * @brief Check if this SyncQueue is empty or not.
+   * @return \c true if SyncQueue is empty, \c false otherwise.
+   */
+  bool Empty() const;
+
+  /**
+   * @brief Closes this SyncQueue, deletes all queued elements and blocks users
+   * from pulling any elements from the queue.
+   */
+  void Close();
+
+  /**
+   * @brief Pulls one element from the SyncQueue.
+   * @param element The element pulled from the queue.
+   * @param closed \c true if this SyncQueue was closed, \c false otherwise.
+   */
+  void Pull(T& element, bool& closed);
```

Comment:
```suggestion
  bool Pull(T& element);
```
Why not return whether Pull method was successful? This will be more convenient, cause user don't have to provide additional variable (bool closed);

Reply:
But what does return true/false mean. You call Pull() and it returns false on success and true on a closed queue. This will be confusing in my opinion. That's why the variable is more self-explanatory and user-friendly.

Reply:
>  That's why the variable is more self-explanatory and user-friendly.

The current approach is not user-friendly as-well, cause user has to declare addition bool variable, even when he has no intention to use at all.

> But what does return true/false mean. 

Returning _false_ when the queue is closed - this means that the operation wasn't successful. True otherwise.

----------------------------
An alternate approach is to use a pointer instead of reference and default-initialize it to nullptr:
```suggestion
  void Pull(T& element, bool* closed = nullptr);
```
But the approach with returning true/false for me seems better.